exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntimeAdapter.java

```diff
@@ -163,25 +163,30 @@ private static ServiceInstanceSpec parseInstanceSpec(byte[] instanceSpec) {
    * Executes the service transaction.
    *
    * @param serviceId the service numeric identifier
+   * @param interfaceName the name of the interface in which the transaction is defined
    * @param txId the transaction type identifier within the service
    * @param arguments the transaction arguments
    * @param forkNativeHandle a handle to a native fork object
+   * @param callerServiceId the id of the service which invoked the transaction (in case of
+   *      inner transactions; or 0 when the caller is an external message
```

Comment:
nit: missing closing brace

Reply:
Also, I can't find any mention that `0` can't be a service ID. Doesn't `-1` suit better in this case?

update: `0` is a supervisor service's ID, but supervisor service isn't strictly essential for blockchain operations (or it shouldn't be, at least).

Reply:
0 here doesn't matter as it isn't supposed to be used in the case, same as hash code and public key in the other case (which are all zeroes). Whether a transaction is external is determined by the interface name currently. It could be set to something different to have something different from the supervisor id, but I hope this signature won't last long.

Reply:
> Whether a transaction is external is determined by the interface name currently.

The keyword is "currently." I'm not 100% sure that we'll keep the inability of external clients to call non-default service interface in 1.0, but this restriction will be lifted _eventually_. In this case the current implementation seemingly becomes _dangerous_: any transaction from the external client will be considered to be authorized by the supervisor! (I'm sure it'll be fixed by that time.)

Reply:
The thing is that what is currently available is very limited and I don't expect this implementation to last (particularly, this monstrous parameter list). It just doesn't support (and explicitly forbids) anything but 'exonum.Configure'. Supporting anything else will require changes.